test/ziggurat/config_test.clj

```diff
@@ -13,69 +13,107 @@
+    (let [config-filename        "config.test.edn"
+          config-values-from-env (-> (config-from-env config-filename)
                                      (update-in [:ziggurat] dissoc :nrepl-server))]
-      (with-redefs [config-from-env (fn [_] config-values-from-env)]
+      (with-redefs [config-from-env (fn [_] config-values-from-env)
+                    config-file     config-filename]
         (mount/start #'config)
         (is (= (-> default-config :ziggurat :nrepl-server :port) (-> config :ziggurat :nrepl-server :port))) ;; when env variable is missing, takes default value
         (is (= (-> config-values-from-env :ziggurat :stream-router :default :bootstrap-servers) (-> config :ziggurat :stream-router :default :bootstrap-servers))) ;; when key is not present in default config, takes env variable
         (is (= (-> config-values-from-env :ziggurat :rabbit-mq :delay :queue-name) (-> config :ziggurat :rabbit-mq :delay :queue-name))) ;; when key is present in both default and env variables, takes env variable
         (mount/stop))))
 
   (testing "returns default interpolated rabbitmq config when not present in env variables"
-    (let [app-name "application_name"
-          config-values-from-env (-> (config-from-env "config.test.edn")
-                                     (update-in [:ziggurat :rabbit-mq] dissoc :delay)
-                                     (assoc-in [:ziggurat :app-name] app-name))
+    (let [app-name                    "application_name"
+          config-filename             "config.test.edn"
+          config-values-from-env      (-> (config-from-env config-filename)
+                                          (update-in [:ziggurat :rabbit-mq] dissoc :delay)
+                                          (assoc-in [:ziggurat :app-name] app-name))
           expected-delay-queue-config {:queue-name           "application_name_delay_queue"
                                        :exchange-name        "application_name_delay_exchange"
                                        :dead-letter-exchange "application_name_instant_exchange"
                                        :queue-timeout-ms     5000}]
-      (with-redefs [config-from-env (fn [_] config-values-from-env)]
+      (with-redefs [config-from-env (fn [_] config-values-from-env)
+                    config-file     config-filename]
         (mount/start #'config)
         (is (= expected-delay-queue-config (-> config :ziggurat :rabbit-mq :delay)))
         (mount/stop)))))
 
 (deftest ziggurat-config-test
   (testing "returns ziggurat config"
-    (let [config-values-from-env (config-from-env "config.test.edn")]
-      (with-redefs [config-from-env (fn [_] config-values-from-env)]
+    (let [config-filename        "config.test.edn"
+          config-values-from-env (config-from-env config-filename)]
+      (with-redefs [config-from-env (fn [_] config-values-from-env)
+                    config-file     config-filename]
```

Comment:
Hey michael, can you help me understand how redefining a config-file here is helping us?

Reply:
I apologize in advance if my explanation is not clear, but I will do my best. The test is getting configurations from 2 different sources, from "config.test.edn" (as provided in the test) and "config.edn" (as per line 12 of config.clj. in our test, config.edn does not exist. but regardless it exists or not, it get merged with the default-config) then you are comparing them (the config you got from config-from-env and the config we got from the function in test, in this case statsd-config). Redefining config-file in the test ensures that it only gets the configuration from 1 source (whichever config file that was provided to the test). There are cases that the configuration (thru the config file that is provided to the test) is different from what we got from the merged config (config.edn + default-config). The best example of this are the test cases for the datadog/statsd.